core/deployment/src/test/java/io/quarkiverse/langchain4j/test/tools/ToolExecutionModelTest.java

```diff
@@ -0,0 +1,340 @@
+import org.junit.jupiter.api.condition.EnabledForJreRange;
+import org.junit.jupiter.api.condition.JRE;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.testcontainers.shaded.org.awaitility.Awaitility;
+
+import dev.langchain4j.agent.tool.Tool;
+import dev.langchain4j.agent.tool.ToolExecutionRequest;
+import dev.langchain4j.agent.tool.ToolSpecification;
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.data.message.ChatMessage;
+import dev.langchain4j.data.message.ToolExecutionResultMessage;
+import dev.langchain4j.memory.ChatMemory;
+import dev.langchain4j.memory.chat.ChatMemoryProvider;
+import dev.langchain4j.model.chat.ChatLanguageModel;
+import dev.langchain4j.model.output.FinishReason;
+import dev.langchain4j.model.output.Response;
+import dev.langchain4j.model.output.TokenUsage;
+import dev.langchain4j.service.MemoryId;
+import dev.langchain4j.service.UserMessage;
+import io.quarkiverse.langchain4j.RegisterAiService;
+import io.quarkiverse.langchain4j.ToolBox;
+import io.quarkiverse.langchain4j.runtime.aiservice.NoopChatMemory;
+import io.quarkiverse.langchain4j.test.Lists;
+import io.quarkus.arc.Arc;
+import io.quarkus.test.QuarkusUnitTest;
+import io.quarkus.virtual.threads.VirtualThreadsRecorder;
+import io.smallrye.common.annotation.NonBlocking;
+import io.smallrye.common.annotation.RunOnVirtualThread;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.Vertx;
+
+public class ToolExecutionModelTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(MyAiService.class, Lists.class));
+
+    @Inject
+    MyAiService aiService;
+
+    @Inject
+    Vertx vertx;
+
+    @Test
+    @ActivateRequestContext
+    void testBlockingToolInvocationFromWorkerThread() {
+        String uuid = UUID.randomUUID().toString();
+        var r = aiService.hello("abc", "hi - " + uuid);
+        assertThat(r).contains(uuid, Thread.currentThread().getName()); // We are invoked on the same thread
+    }
+
+    @Test
+    @ActivateRequestContext
+    void testBlockingToolInvocationFromEventLoop() {
+        String uuid = UUID.randomUUID().toString();
+        AtomicReference<Throwable> failure = new AtomicReference<>();
+        vertx.getOrCreateContext().runOnContext(x -> {
+            try {
+                Arc.container().requestContext().activate();
+                aiService.hello("abc", "hi - " + uuid);
+            } catch (IllegalStateException e) {
+                failure.set(e);
+            } finally {
+                Arc.container().requestContext().deactivate();
+            }
+        });
+
+        Awaitility.await().until(() -> failure.get() != null);
+        assertThat(failure.get()).hasMessageContaining("Cannot execute blocking tools on event loop thread");
+    }
+
+    @Test
+    @ActivateRequestContext
+    @EnabledForJreRange(min = JRE.JAVA_21)
```

Comment:
+1

Reply:
that's the beauty of our model - it does not have a dependency on the JDK 21 API :-)